keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md

```diff
@@ -0,0 +1,221 @@
+
+- `platforms/`
+  - `darwin/`, `linux/`, `windows/`
+    - `amd64/`, `arm/`, `arm64/`, `ppc64le/`, `s390x/`
+
+For comparison, the existing platform-specific tarballs
+(such as `kubernetes-client-linux-amd64.tar.gz`) place all binaries under
+a constant path with no platform information: `kubernetes/client/bin/kubectl`.
+
+Scripts (such as `cluster/get-kube-binaries.sh`) [extract these tarballs
+back into platform-specific directories](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/cluster/get-kube-binaries.sh#L143-L156)
+to support downloading multiple platforms into a single workspace.
+
+The test tarball should follow the lead of the other platform-specific tarballs
+and place the binaries under `test/bin`. We can then reuse the existing
+functionality already implemented for the other tarballs.
+
+#### Portable sources
+
+Portable sources are basically copied directly from the source tree:
+
+- `test/e2e/testing-manifests/`
+- `test/images/`
+- `test/kubemark/`
+- `hack/` ([partially](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/hack/lib/golang.sh#L193-L197))
+
+We have two options for these:
+
+1. Continue to distribute as a separate tarball, either `kubernetes-test.tar.gz`,
+   or possibly something like `kubernetes-test-portable.tar.gz`.
+- This makes the distinction very clear vs. the binary artifacts
+- There's already some precedent, such as the `kubernetes-manifest.tar.gz` tarball
+- It slightly complicates downloading of test dependencies
+2. Duplicate these sources into each binary-specific tarball.
+- Simplifies test dependency distribution
+- Portable sources are small (as of v1.13.2, approximately 2.7MB uncompressed
+  or about 186KB compressed) so duplication isn't a huge worry
+- Complicates extraction of tarballs with existing scripts, since they assume
+  everything is platform-specific
+
+It's not immediately obvious which of these is the better approach based on the
+tradeoffs.
+
+### Updating dependencies on `kubernetes-test.tar.gz`
+
+Currently the CI workflows and `kubetest` use the `cluster/get-kube.sh` and
+`cluster/get-kube-binaries.sh` scripts to download all artifacts, and
+conveniently `get-kube-binaries.sh` is versioned with the release artifacts in
+`kubernetes.tar.gz`, so simply making `get-kube-binaries.sh` aware of the new
+tarballs should be sufficient for most CI and developer needs.
+
+Because the test tarball includes binaries used both on the host running tests
+(such as the `e2e` binary), as well as binaries which may run the nodes
+(`e2e.node`), we would need to make sure to download binary test artifacts
+targeting the host platform, node platform, and possibly server platform.
+
+A quick search reveals a few other uses of `kubernetes-test.tar.gz`, mostly in
+`cluster/`. We can update these to use the platform-specific tarballs, possibly
+with a fallback to the mondo-tarball if worried about versioning.
```

Comment:
This is making me think about graduation criteria.  Is the goal to _not_ be building a mondo-tarball by the time we release 1.14.0?  Or provide both mondo and split for some time?

Reply:
my preference would be a clean break, no mondo-tarball by 1.14.0.

Reply:
I added a bit about this into the "risks/mitigations" section, though I'm not sure if it'd be more appropriate somewhere else.

Reply:
Hi @ixdy,

I think this is a public part of the workflow and thus should be deprecated in accordance with any other public feature -- two releases. The mondo tarball should continue to exist in 1.14 and 1.15 with a deprecation `NOTICE` or `DEPRECATION` file placed inside the tarball and in SIG-Testing's GitHub `README`.